### DISCUSSION TITLE
User's mental model of the Canvas

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-manifest-editor#29] [Ideas]
Some users won't have any notion of the Canvas. The manifest is a sequence of images, one after the other. They can get their job done by assembling a sequence of images. And perhaps, some video or audio too.

Introduce the Canvas when it's required, but don't make the user think about the **space** before they think about the **content** - while a real use case for some, it's not the most common.

This is where the PowerPoint analogy sheds some light. Although it's a good analogy (slide = canvas, put your content on the slide), in practice it doesn't quite match typical usage.

* In PowerPoint, I really am thinking about slides. What am I going to put on the next slide? Add a new slide, type some text, maybe add a diagram. All slides are the same size, unless I go out of my way to change that. I can, however, change the default aspect ratio (e.g., if I know I'm going to be presenting on a 4:3 display).
* In a manifest, I'm usually thinking about the next _image_. I want to start the manifest by adding an image, and continue by adding another. Often what I really want to do is add an image _service_; I want to add another deep zoom to the sequence of deep zooms, or take a deep-zoom from someone else's manifest and put it in mine. 

Faced with the blank slate, or the _next_ operation, it's much more likely my _next_ is a content choice, usually an image, and in fact usually an image _service_. The most common job is _add another image service to the manifest_.

If the manifest editor is being used in a teaching scenario where the exercise is to learn what IIIF is by creating things in a visual environment, then the Canvas **_does_** need to be a part of the model presented by the UI. And users who do know what IIIF is, and come to the manifest editor with the intention of making a layout of multiple assets on the same canvas (e.g., Delft, storytelling), or stacking up a multispectral image set, need to be able to do that clearly without the application hiding things from them.

The UI has to cater for the simple path, but allow the more complex; if it can use the idea of the Canvas to help the user get their job done _when that idea is really a help_, then the didactic purpose might happen more naturally, too.

So what does this mean for the user interface?

Introduce control of canvas dimensions only when required - the idea of Canvas as a layout space can be present without forcing the user to think about the numerical dimensions of that space.

New... (new what?)
or maybe Add... (need to separate adding something new to the current canvas from adding another canvas).

It's OK for this to say "New Canvas" if visually in the UI it's the obvious "thing to press to make a new image" - it could say "New blahblah" and still obviously be the same call to action. This helps establish the terminology, it doesn't matter what it's called, but the app called it a Canvas so let's go with that.

The point at which you add another Canvas:

* From content
     - Browse IIIF (launches #5 and remembers where you were) 
     - Paste URL (be clever about analysing the URL, see below)
* Empty Canvas
     - Show width and height input boxes defaulting to, say, 10000 x 10000

Adding a new piece of content as a new canvas is the happy path.
You can also add a canvas from another manifest.
The pasted URL is analysed to see what it is - a static image? an image service (be clever about detecting this)? If it's a IIIF resource, deref it... if it's a canvas (unlikely) then add right in, if it's a manifest then it's the same as #5, allow selection of a Canvas within the manifest - but only if there's more than one canvas, otherwise just use it).

If all you do is add some images one after the other, you'll get to the end of your created manifest without ever having to interact with UI that requires information about the canvas.

Visually, the editor just shows the image. There's no "canvas" - empty slide - visible. Here the pale grey is the backdrop of the editor.

![hunters1](https://user-images.githubusercontent.com/1443575/143880889-bcc70961-ba67-42ac-a567-f2cb2d8064ee.png)

You can't move this around on the canvas - it _is_ the canvas, we're not (by default) allowing it to be separately edited.

Even if you added another image to this canvas, you still wouldn't necessarily see the canvas - e.g., if the new image was wholly enclosed by the original:

![chateauroux](https://user-images.githubusercontent.com/1443575/143881010-ccc2dbae-1e36-4766-84e7-d4a6492f7398.png)

If I dragged this image to the side, or added another image, we'd then start to see the _bounding box_ canvas make its first appearance:

![hunters2](https://user-images.githubusercontent.com/1443575/143881127-451f0d60-b5fe-41e0-91c0-08f1e7308f9c.png)

This is the contentious bit...

_Unlike_ a more traditional drawing tool, the "paper" that you've started to see here isn't fixed - it will still default to the bounding box. It changes shape as you move things around, always snapping to the bounding box. 

That is, unless you take control of the canvas yourself by clicking a button to do so ("lock canvas"? "unlock canvas"?).

This gives you access to drag handles on the canvas borders, and also a width and height input box for the canvas. This updates as you move the canvas borders, and the canvas borders will move if you enter different values in the input boxes.

![hunters3](https://user-images.githubusercontent.com/1443575/143881397-4b8c769e-1ea6-499e-b543-5eb76b1ea6c0.png)

The starting values for these input boxes are whatever the implied canvas dimensions were at the point the canvas was unlocked.
This is based on the images on the canvas, subject to a minimum.

In the "New" UI mentioned earlier, you'd get an unlocked canvas by default, as a default size, only if you selected "Empty Canvas".

That graph paper background could be something else (I borrowed it from Lucidchart) but whatever it is, it's distinct from the application window background (the pale grey). And crucially it won't appear if all you are doing is adding single images in sequence to a manifest, or importing canvases from other places. It only comes into view if you:

 - have more than one piece of content on the canvas and they are disjoint, or overlap but one is not contained by the other
 - start with an empty canvas

As far as possible the user never need set canvas dimensions, even in the above example.
Unlike (say) Lucidchart, the visible graph paper isn't there to give an indication of absolute size - what's going to happen if you print the thing. It's purely for establishing a coordinate space with no notion of absolute size. So it can borrow some behaviour from tools like Lucidchart, but not go too far.

**Comment by tomcrane** [comment]
Technical aside - some idea of what might be happening with Canvas Panel as the editing surface can be seen in https://deploy-preview-50--iiif-canvas-panel.netlify.app/docs/examples/canvas-on-canvas - the `paint` and `setDimensionsFromContent()` functions.

Is the graph-paper a lowest-z-index layer in the _editing-mode_ canvas? The background of canvas-panel's container showing through? What's happening when you resize the canvas grab handles?

**Comment by stephenwf** [comment]
> If I dragged this image to the side, or added another image, we'd then start to see the bounding box canvas make its first appearance:

From our conversation on slack, I think this should be a much more intentional operation from the user. A conversion from a simple, compatible canvas, to a composition. This will help us when it comes to rendering the editor itself, optimising for the common simple case, but also will help avoid users accidentally creating an incompatible canvas.

**Comment by sammeltassen** [comment]
I would like to be able to set (and lock) the canvas ratio rather than canvas dimensions, with optionally some standard options:

- 1:1
- 1:1.41421 (paper size)
- 1:2
- 2:3
- 3:4
- 4:5
- 9:16 (screen dimension)

Also reversible...

UX wise I can imagine showing the canvas even in the single image scenario, as a kind of shadow or glow around it.

Shouldn't there just be an option to show/hide the canvas, even with a single image?

**Comment by sammeltassen** [comment]
The Faculty of Architecture at TU Delft is currently preparing a visual timeline of their history. I'm involved for the digital counterpart.

This is their mental model:

![tumblr_nzojl08OlN1qe0nlvo1_1280](https://user-images.githubusercontent.com/747644/146408273-ea41c9d6-d7b4-481b-9228-638fc20bbeed.jpg)

Do we imagine making something like this in the Manifest Editor, as a single organised canvas, with labels and annotations?

This relates to the use cases mentioned here:

- https://github.com/digirati-co-uk/iiif-manifest-editor/issues/17#issuecomment-957764328

**Reply by sammeltassen** [reply]
Another example if you'll allow me to indulge:

![IMG_5694-cropped](https://user-images.githubusercontent.com/747644/146414784-5833a6de-b8d3-43fa-952a-6f9120bd50af.jpeg)

In my anachronistic reading, each side of a single wooden stand is a canvas in this exhibition (with a certain proportion), containing multiple images, and the manifest is composed of all the stands.

It's pretty close to IIIF in my view as it's an exhibition of copies, cropped and scaled from the originals, and this is the kind of copyright-free stuff mostly available with IIIF...

**Reply by sammeltassen** [reply]
Last one and then I'll stop going on about this: Hockney's [The Great Wall](https://www.thedavidhockneyfoundation.org/artwork/3376) could also be a IIIF-project.

**Reply by sammeltassen** [reply]
So perhaps there're three different experiences/mental models:

- Single
- Layered
- Composition

Each one optionally complemented with text annotations...

**Reply by sammeltassen** [reply]
Elaboration:

- Single image, with an emphasis on the thumbnail row; no need for a visible canvas or layers
- Layered image, with an emphasis on the layer panel for reordering (and opacity changes?); no need for a visible canvas or thumbnail row
- Visual composition, with an emphasis on the canvas itself; no need for a layer panel (if no overlap) and thumbnail row

Of course in reality more nuanced, but maybe conceptually useful.

**Reply by sammeltassen** [reply]
Further elaboration. Three interface levels:

1) Manifest-mode
- Contact sheet
- Reordering images
- Editing manifest label, summary and metadata
- Importing/exporting manifests

2) Canvas-mode
- Image and thumbnail row
- No canvas editing
- Editing canvas label, summary and metadata
- Importing/exporting canvasses

3) Annotation-mode
- Thumbnail row hidden
- Layers panel visible
- Image manipulation tools
- Drawing, scaling and moving annotations
- Editing annotation labels, summary and metadata
- Importing/exporting annotations

A bit like the different steps in the [Allmaps Editor](https://editor.allmaps.org/#/).

This is both educative (to become aware of the different manifest levels) and useful :)

**Reply by tomcrane** [reply]
In which one of these three levels would you perform the action of creating a new canvas by adding a new image to the manifest?

In theory you are editing a canvas (both creating the canvas and putting the image on it) but in the vast majority of cases, the "mental model" is simply adding a new image to the sequence of images (and the manifest editor creates a canvas for it quietly in the background). This could also happen in modes 1 and 2 without any canvas concept intruding.

**Reply by sammeltassen** [reply]
I'd do this in the contact sheet view. Two main views could be sufficient:

1. Contact Sheet (manifest view)

- Arranging canvasses
- Adding/removing canvasses
- Editing manifest metadata
- Editing ranges, ids...

2. Single image (canvas view)

- Arranging annotations (position and layers)
- Adding/removing annotations
- Editing canvas metadata
- Editing annotation metadata > this can still be confusing side by side to canvas metadata, perhaps a sub-view, similar to when you open up a group of objects in Illustrator

Perhaps I'm confused by 'mental model'. Don't we want the users to become more aware of the different concepts of IIIF? Can't the Manifest Editor partly shape their mental model? Shouldn't the editor encourage the making of more complex canvasses with multiple annotations? I hope some of the more advanced options can also be visible in the simple use-cases, without disturbing the overall experience.
